daemon/libnetwork/portmappers/nat/mapper_linux.go

```diff
@@ -74,28 +70,45 @@ func (pm PortMapper) MapPorts(ctx context.Context, cfg []portmapperapi.PortBindi
 		}
 	}
 
-	// Try up to maxAllocatePortAttempts times to get a port that's not already allocated.
-	var bindings []portmapperapi.PortBinding
-	var err error
-	for i := 0; i < maxAllocatePortAttempts; i++ {
-		bindings, err = pm.attemptBindHostPorts(ctx, cfg, proto, hostPort, hostPortEnd, fwn)
-		if err == nil {
-			break
-		}
-		// There is no point in immediately retrying to map an explicitly chosen port.
-		if hostPort != 0 && hostPort == hostPortEnd {
-			log.G(ctx).WithError(err).Warnf("Failed to allocate and map port")
-			return nil, err
+	bindings := make([]portmapperapi.PortBinding, 0, len(cfg))
+	defer func() {
+		if retErr != nil {
+			if err := pm.UnmapPorts(ctx, bindings, fwn); err != nil {
+				log.G(ctx).WithFields(log.Fields{
+					"pbs":   bindings,
+					"error": err,
+				}).Warn("Failed to release port bindings")
+			}
 		}
-		log.G(ctx).WithFields(log.Fields{
-			"error":   err,
-			"attempt": i + 1,
-		}).Warn("Failed to allocate and map port")
+	}()
+
+	addrs := make([]net.IP, 0, len(cfg))
+	for i := range cfg {
+		cfg[i] = setChildHostIP(pm.pdc, cfg[i])
+		addrs = append(addrs, cfg[i].ChildHostIP)
 	}
 
+	pa := portallocator.NewOSAllocator()
+	allocatedPort, socks, err := pa.RequestPortsInRange(addrs, proto, int(hostPort), int(hostPortEnd))
 	if err != nil {
-		// If the retry budget is exhausted and no free port could be found, return
-		// the latest error.
+		return nil, err
+	}
+
+	for i := range cfg {
+		pb := portmapperapi.PortBinding{
+			PortBinding: cfg[i].PortBinding.Copy(),
+			BoundSocket: socks[i],
+			ChildHostIP: cfg[i].ChildHostIP,
```

Comment:
This one is OK to be copied as-is, or do we need to change it to something like;

```
			ChildHostIP: slices.Clone(cfg[i].ChildHostIP),
```